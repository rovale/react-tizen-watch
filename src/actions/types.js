export const INITIALIZE_APP = 'INITIALIZE_APP';

export const LOAD_PAGES = 'LOAD_PAGES';
export const LOAD_DEVICES = 'LOAD_DEVICES';
export const LOAD_GROUPS = 'LOAD_GROUPS';

export const ACTIVATE_PAGE = 'ACTIVATE_PAGE';
export const SELECT_PAGE = 'SELECT_PAGE';
export const CLOSE_PAGE = 'CLOSE_PAGE';

export const ACTIVATE_DEVICE = 'ACTIVATE_DEVICE';
export const SELECT_DEVICE = 'SELECT_DEVICE';
export const CLOSE_DEVICE = 'CLOSE_DEVICE';
export const CHANGE_DEVICE_ATTRIBUTE = 'CHANGE_DEVICE_ATTRIBUTE';
export const TOGGLE_SWITCH = 'TOGGLE_SWITCH';

export const ACTIVATE_BUTTON = 'ACTIVATE_BUTTON';
export const CLICK_BUTTON = 'CLICK_BUTTON';
