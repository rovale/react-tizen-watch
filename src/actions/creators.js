import * as actionType from './types';

export const initializeApp = () => ({ type: actionType.INITIALIZE_APP });
export const handleError = (msg, url, lineNo, columnNo, error) => ({ type: actionType.HANDLE_ERROR, payload: { msg, url, lineNo, columnNo, error } });
export const loadPages = pages => ({ type: actionType.LOAD_PAGES, payload: { pages } });
export const loadDevices = devices => ({ type: actionType.LOAD_DEVICES, payload: { devices } });

export const activatePage = id => ({ type: actionType.ACTIVATE_PAGE, payload: { id } });
export const selectPage = id => ({ type: actionType.SELECT_PAGE, payload: { id } });
export const closePage = () => ({ type: actionType.CLOSE_PAGE });

export const activateDevice = id => ({ type: actionType.ACTIVATE_DEVICE, payload: { id } });
export const selectDevice = id => ({ type: actionType.SELECT_DEVICE, payload: { id } });
export const closeDevice = () => ({ type: actionType.CLOSE_DEVICE });
export const changeDeviceAttribute = event => ({ type: actionType.CHANGE_DEVICE_ATTRIBUTE, payload: { event } });

export const toggleSwitch = id => ({ type: actionType.TOGGLE_SWITCH, payload: { id } });

export const activateButton = id => ({ type: actionType.ACTIVATE_BUTTON, payload: { id } });
export const clickButton = id => ({ type: actionType.CLICK_BUTTON, payload: { id } });
