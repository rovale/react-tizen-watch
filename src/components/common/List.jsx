import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

export class List extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    activeOptionId: PropTypes.string.isRequired,
    onActivateOption: PropTypes.func.isRequired,
    onSelectOption: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    this.onRotaryDetent = this.onRotaryDetent.bind(this);
    this.onSelectItem = this.onSelectItem.bind(this);
  }

  componentDidMount() {
    window.addEventListener('rotarydetent', this.onRotaryDetent);
  }

  componentWillUnmount() {
    window.removeEventListener('rotarydetent', this.onRotaryDetent);
  }

  onRotaryDetent(e) {
    const { direction } = e.detail;

    const index = this.props.options.findIndex(item => item.id === this.props.activeOptionId);

    if (direction === 'CW') {
      if (index < (this.props.options.length - 1)) {
        this.props.onActivateOption(this.props.options[index + 1].id);
      }
    } else if (index > 0) {
      this.props.onActivateOption(this.props.options[index - 1].id);
    }
  }

  onSelectItem(itemId) {
    this.props.onSelectOption(itemId);
  }

  render() {
    const { options } = this.props;

    const items = options.map(item => (
      <Item
        key={item.id}
        id={item.id}
        active={item.id === this.props.activeOptionId}
        onSelect={item.onSelect || this.onSelectItem}
      >
        {item.content || <div>{item.title}</div>}
      </Item>
    ));

    return (
      <ul className="ui-listview ui-snap-listview">
        {items}
      </ul>
    );
  }
}

export class Item extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
  };

  static contextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.scrollToMiddle = this.scrollToMiddle.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.element = null;
  }

  componentDidMount() {
    if (this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(1), 0);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.active && !this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(400), 0);
    }
  }

  onSelect(e) {
    e.preventDefault();
    this.props.onSelect(this.props.id);
  }

  scrollToMiddle(duration) {
    const itemRect = this.element.getBoundingClientRect();
    this.context.scrollToMiddle(duration, itemRect.top, itemRect.height);
  }

  render() {
    const { id, active, children } = this.props;
    return (
      <li ref={(e) => { this.element = e; }} className={`ui-snap-listview-item${active ? ' ui-snap-listview-selected' : ''}`}>
        <a href={`#${id}`} onClick={this.onSelect} id={id}>{children}</a>
      </li>
    );
  }
}
