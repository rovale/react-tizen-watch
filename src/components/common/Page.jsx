import React from 'react';
import PropTypes from 'prop-types';
import scroll from 'smooth-move';

export class Page extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  static childContextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.element = null;
  }

  getChildContext() {
    return { scrollToMiddle: this.scrollToMiddle };
  }

  scrollToMiddle = (duration, itemTop, itemHeight) => {
    const scroller = this.element.getElementsByClassName('ui-scroller')[0];
    const scrollerHeight = scroller.getBoundingClientRect().height;
    const newScrollTop = (itemTop + scroller.scrollTop) - ((scrollerHeight - itemHeight) / 2);

    scroll(scroller, { x: 0, y: newScrollTop, duration });
  }

  render() {
    return (
      <div ref={(e) => { this.element = e; }} className="ui-page ui-page-active ui-scroll-on">
        <div className="ui-scroller ui-snap-container" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

export const Header = ({ children }) => (
  <header className="ui-header">
    <h2 className="ui-title">{children}</h2>
  </header>
);

Header.propTypes = {
  children: PropTypes.string.isRequired,
};

export const Content = ({ children }) => {
  return (
    <div className="ui-content" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
      {children}
    </div>);
};

Content.propTypes = {
  children: PropTypes.node.isRequired,
};
