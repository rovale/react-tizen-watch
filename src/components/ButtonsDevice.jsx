import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { List } from './common/List';

import * as action from '../actions/creators';

let ButtonsDevice = ({ device, activeButtonId, onActivateButton, onClickButton }) => {
  const buttons = device.config.buttons;
  const activeOptionId = activeButtonId || buttons[0].id;

  return (<List
    options={buttons.map(b => ({ id: b.id, title: b.text }))}
    activeOptionId={activeOptionId}
    onActivateOption={onActivateButton}
    onSelectOption={onClickButton}
  />);
};

ButtonsDevice.propTypes = {
  device: PropTypes.object.isRequired,
  activeButtonId: PropTypes.string,
  onActivateButton: PropTypes.func.isRequired,
  onClickButton: PropTypes.func.isRequired,
};

ButtonsDevice.defaultProps = {
  activeButtonId: null,
};

const mapStateToProps = state => ({
  activeButtonId: state.ui.activeButtonId,
});

const mapDispatchToProps = dispatch => ({
  onActivateButton: id => dispatch(action.activateButton(id)),
  onClickButton: id => dispatch(action.clickButton(id)),
});

ButtonsDevice = connect(mapStateToProps, mapDispatchToProps)(ButtonsDevice);

export default ButtonsDevice;
