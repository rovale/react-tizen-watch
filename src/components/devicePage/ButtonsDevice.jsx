import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { List } from '../common/List';

import * as action from '../../actions/creators';

const ButtonsDevice = ({
  device, activeButtonId, onActivateButton, onClickButton,
}) => {
  const { buttons } = device.config;
  const activeOptionId = activeButtonId || buttons[0].id;

  const getListItemContent = button => (
    <div>
      <div className="vertical-align-middle">
        <input
          type="radio"
          readOnly
          checked={button.id === device.attributes[0].value}
        />
      </div>
      <div className="vertical-align-middle">
        &nbsp;{button.text}
      </div>
    </div>);

  return (<List
    options={buttons.map(b => ({
      id: b.id,
      content: getListItemContent(b),
    }))}
    activeOptionId={activeOptionId}
    onActivateOption={onActivateButton}
    onSelectOption={onClickButton}
  />);
};

ButtonsDevice.propTypes = {
  device: PropTypes.shape({
    attributes: PropTypes.array.isRequired,
    config: PropTypes.object.isRequired,
  }).isRequired,
  activeButtonId: PropTypes.string,
  onActivateButton: PropTypes.func.isRequired,
  onClickButton: PropTypes.func.isRequired,
};

ButtonsDevice.defaultProps = {
  activeButtonId: null,
};

const mapStateToProps = state => ({
  activeButtonId: state.ui.activeButtonId,
});

const mapDispatchToProps = dispatch => ({
  onActivateButton: id => dispatch(action.activateButton(id)),
  onClickButton: id => dispatch(action.clickButton(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ButtonsDevice);
