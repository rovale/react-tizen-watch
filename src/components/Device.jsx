import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Page as UiPage, Header, Content } from './common/Page';
import ButtonsDevice from './ButtonsDevice';
import * as action from '../actions/creators';

let Device = ({ device, onClose }) => {
  const getDeviceContent = () => {
    switch (device.template) {
      case 'buttons':
        return <ButtonsDevice device={device} />;
      default:
        return <div>Content of {device.id} {device.template}</div>;
    }
  };

  return (<UiPage>
    <Header>{device.title}</Header>
    <Content>
      {getDeviceContent(device)}
    </Content>
    {/* <footer className="ui-footer ui-bottom-button ui-fixed">
      <a href="#close" onClick={onClose} className="ui-btn">OK</a>
    </footer> */}
  </UiPage>);
};

Device.propTypes = {
  device: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  device: state.devices.find(d => d.id === state.ui.selectedDeviceId),
});

const mapDispatchToProps = dispatch => ({
  onClose: () => dispatch(action.closeDevice()),
});

Device = connect(mapStateToProps, mapDispatchToProps)(Device);

export default Device;
