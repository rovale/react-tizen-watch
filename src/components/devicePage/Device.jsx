import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Page as UiPage, Header, Content } from '../common/Page';
import ButtonsDevice from './ButtonsDevice';

const Device = ({ device }) => {
  const getDeviceContent = () => {
    switch (device.template) {
      case 'buttons':
        return <ButtonsDevice device={device} />;
      default:
        return <div>Content of {device.id} {device.template}</div>;
    }
  };

  return (
    <UiPage>
      <Header>{device.title}</Header>
      <Content>
        {getDeviceContent(device)}
      </Content>
    </UiPage>
  );
};

Device.propTypes = {
  device: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  device: state.devices.find(d => d.id === state.ui.selectedDeviceId),
});

export default connect(mapStateToProps)(Device);
