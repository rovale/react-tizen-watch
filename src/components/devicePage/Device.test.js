import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { createStore } from 'redux';
import reducer from '../../reducers';
import * as action from '../../actions/creators';

import Device from './Device';
import ButtonsDevice from './ButtonsDevice';
import { List, Item } from './../common/List';

Enzyme.configure({ adapter: new Adapter() });

/* eslint-disable object-curly-newline */
const devices = [
  { id: 'someDisplayDevice', name: 'SDD', template: 'device', attributes: [] },
  { id: 'somePresenceDevice', name: 'SPD', template: 'presence', attributes: [] },
  { id: 'someSwitch', name: 'SS', template: 'switch', attributes: [] },
  { id: 'someTemperatureDevice', name: 'STD', template: 'temperature', attributes: [] },
  { id: 'someThermostat', name: 'ST', template: 'thermostat', attributes: [] },
  { id: 'someButtons', name: 'SB', template: 'buttons', attributes: [] },
];
/* eslint-enable object-curly-newline */

describe('The Device component', () => {
  let store;

  beforeEach(() => {
    store = createStore(reducer);
    store.dispatch(action.loadDevices(devices));
  });

  test('it should show a ButtonsDevice component for a buttons device', () => {
    store.dispatch(action.selectDevice('someButtons'));
    const device = shallow((<Device store={store} />)).dive();

    const buttonsDevice = device.find(ButtonsDevice);

    expect(buttonsDevice).toHaveLength(1);
  });

  test('it should show some text for a device without a specific implementation', () => {
    store.dispatch(action.selectDevice('someTemperatureDevice'));
    const device = shallow((<Device store={store} />)).dive();


    expect(device.dive()).toMatchSnapshot();
  });
});
