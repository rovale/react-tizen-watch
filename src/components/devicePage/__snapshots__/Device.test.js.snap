// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Device component it should show some text for a device without a specific implementation 1`] = `
<div
  className="ui-page ui-page-active ui-scroll-on"
>
  <div
    className="ui-scroller ui-snap-container"
  >
    <Component>
      STD
    </Component>
    <Component>
      <div>
        Content of 
        someTemperatureDevice
         
        temperature
      </div>
    </Component>
  </div>
</div>
`;
